File: plugins/datasources/yesoreyeram-teable-datasource/pkg/plugin/filter.go
package plugin

import (
	"strconv"
	"strings"
)

// FilterNode is a node in the structured filter tree sent from the query editor.
// A node is either a condition (kind="condition") or a group (kind="group").
//
// Teable filters are JSON objects, not query strings. Each condition compiles to
// {"fieldId": <name>, "operator": <op>, "value": <coerced>} and groups compile
// to {"conjunction": "and"|"or", "filterSet": [...]}. The condition needs the
// field's logical category (text/number/boolean/date/select/multiSelect) to
// shape and coerce the value.
type FilterNode struct {
	Kind string `json:"kind"`

	// Condition fields.
	// Field is the Teable field NAME (filters are sent with fieldKeyType=name).
	Field string `json:"field,omitempty"`
	// Category is the logical field category, used to coerce/shape the value.
	Category string `json:"category,omitempty"`
	// Op is the Teable filter operator (is, isNot, contains, isGreater, ...).
	Op    string `json:"op,omitempty"`
	Value string `json:"value,omitempty"`

	// Group fields.
	Connector string       `json:"connector,omitempty"`
	Children  []FilterNode `json:"children,omitempty"`
}

// SortItem is a single sort directive: a field name and a direction.
type SortItem struct {
	Field     string `json:"field"`
	Direction string `json:"direction"`
}

// filterItem is a single Teable filter condition.
type filterItem struct {
	FieldID  string `json:"fieldId"`
	Operator string `json:"operator"`
	Value    any    `json:"value"`
}

// filterSet is a Teable filter group: {conjunction, filterSet:[item|group, ...]}.
// The top-level filter is always a filterSet.
type filterSet struct {
	Conjunction string `json:"conjunction"`
	FilterSet   []any  `json:"filterSet"`
}

// Operator arity classifications.
const (
	arityNone   = "none"   // unary, no value (isEmpty / isNotEmpty) -> value null
	arityList   = "list"   // value is an array (isAnyOf / hasAnyOf / ...)
	aritySingle = "single" // value is a single scalar / date object
)

// BuildFilter converts a structured filter tree (the root group) into a Teable
// filter object ({conjunction, filterSet}) suitable for the `filter` query
// parameter. It returns nil when there are no valid conditions, so the caller
// can omit the filter entirely.
func BuildFilter(root *FilterNode) *filterSet {
	if root == nil {
		return nil
	}
	return compileGroup(root)
}

func compileGroup(group *FilterNode) *filterSet {
	parts := make([]any, 0, len(group.Children))
	for i := range group.Children {
		child := group.Children[i]
		if child.Kind == "group" {
			if g := compileGroup(&child); g != nil {
				parts = append(parts, g)
			}
			continue
		}
		if c := compileCondition(child); c != nil {
			parts = append(parts, *c)
		}
	}
	if len(parts) == 0 {
		return nil
	}
	conjunction := "and"
	if strings.EqualFold(group.Connector, "or") {
		conjunction = "or"
	}
	return &filterSet{Conjunction: conjunction, FilterSet: parts}
}

func compileCondition(c FilterNode) *filterItem {
	field := strings.TrimSpace(c.Field)
	if field == "" {
		return nil
	}
	op := strings.TrimSpace(c.Op)
	if op == "" {
		op = defaultOperator(c.Category)
	}

	switch operatorArity(op) {
	case arityNone:
		// Unary operators must carry a null value per the Teable schema.
		return &filterItem{FieldID: field, Operator: op, Value: nil}
	case arityList:
		values := splitList(c.Value)
		if len(values) == 0 {
			return nil
		}
		return &filterItem{FieldID: field, Operator: op, Value: values}
	default: // single
		v := strings.TrimSpace(c.Value)
		if v == "" {
			return nil
		}
		return &filterItem{FieldID: field, Operator: op, Value: coerceValue(c.Category, v)}
	}
}

// operatorArity classifies how many values a Teable operator consumes.
func operatorArity(op string) string {
	switch op {
	case "isEmpty", "isNotEmpty":
		return arityNone
	case "isAnyOf", "isNoneOf", "hasAnyOf", "hasAllOf", "hasNoneOf", "isExactly", "isNotExactly":
		return arityList
	default:
		return aritySingle
	}
}

// defaultOperator returns a sensible default operator when none is given.
func defaultOperator(category string) string {
	if category == "multiSelect" {
		return "hasAnyOf"
	}
	return "is"
}

// coerceValue converts the string value from the editor into the JSON type
// Teable expects for the given category. Numbers become float64, checkbox
// becomes bool, dates become a {mode:exactDate,...} object; everything else
// stays a string.
func coerceValue(category, value string) any {
	switch category {
	case "number":
		if f, err := strconv.ParseFloat(strings.TrimSpace(value), 64); err == nil {
			return f
		}
		return value
	case "boolean":
		return parseBool(value)
	case "date":
		// Teable date operators expect a structured value with a mode and a
		// timezone. exactDate compares against a specific ISO date.
		return map[string]any{
			"mode":      "exactDate",
			"exactDate": value,
			"timeZone":  "UTC",
		}
	default:
		return value
	}
}

func parseBool(value string) bool {
	switch strings.TrimSpace(strings.ToLower(value)) {
	case "1", "true", "yes", "checked":
		return true
	default:
		return false
	}
}

// splitList splits a comma-separated value into trimmed, non-empty tokens for
// list operators (isAnyOf, hasAnyOf, ...).
func splitList(value string) []string {
	out := make([]string, 0)
	for _, t := range strings.Split(value, ",") {
		t = strings.TrimSpace(t)
		if t != "" {
			out = append(out, t)
		}
	}
	return out
}

File: plugins/datasources/yesoreyeram-teable-datasource/src/filter.ts
import { SelectableValue } from '@grafana/data';

// ---------------------------------------------------------------------------
// Filter model
//
// The structured filter tree is compiled into a Teable JSON `filter` object
// ({conjunction, filterSet:[...]}) on the backend. The frontend only describes
// conditions/groups; it does not build the filter object itself. Operators are
// Teable's native operators and field references use field NAMES (the backend
// queries Teable with fieldKeyType=name).
// ---------------------------------------------------------------------------

export type LogicalConnector = 'and' | 'or';

/** Logical categories a Teable field type can map to for operator selection. */
export type FieldCategory = 'text' | 'number' | 'boolean' | 'date' | 'select' | 'multiSelect' | 'attachment';

export interface FilterCondition {
  kind: 'condition';
  /** Teable field name. */
  field: string;
  /** Logical category of the field; drives backend value coercion/shaping. */
  category?: FieldCategory;
  /** Teable operator (see OPERATORS). */
  op: string;
  /** Comparison value. Ignored for unary operators; comma-separated for lists. */
  value?: string;
}

export interface FilterGroup {
  kind: 'group';
  /** How the direct children are combined (AND/OR). */
  connector: LogicalConnector;
  children: FilterNode[];
}

export type FilterNode = FilterCondition | FilterGroup;

export function newCondition(): FilterCondition {
  return { kind: 'condition', field: '', category: 'text', op: 'is', value: '' };
}

export function newGroup(): FilterGroup {
  return { kind: 'group', connector: 'and', children: [] };
}

export function emptyRootGroup(): FilterGroup {
  return newGroup();
}

// ---------------------------------------------------------------------------
// Operator catalog (field-type aware) — Teable native operators
// ---------------------------------------------------------------------------

export interface OperatorDef {
  value: string;
  label: string;
  /**
   * 'none'   = unary (no value)
   * 'single' = one value
   * 'list'   = comma-separated values compiled into an array
   */
  arity: 'none' | 'single' | 'list';
}

const TEXT: OperatorDef[] = [
  { value: 'is', label: 'is', arity: 'single' },
  { value: 'isNot', label: 'is not', arity: 'single' },
  { value: 'contains', label: 'contains', arity: 'single' },
  { value: 'doesNotContain', label: 'does not contain', arity: 'single' },
];

const NUMBER: OperatorDef[] = [
  { value: 'is', label: '=', arity: 'single' },
  { value: 'isNot', label: '!=', arity: 'single' },
  { value: 'isGreater', label: '>', arity: 'single' },
  { value: 'isGreaterEqual', label: '>=', arity: 'single' },
  { value: 'isLess', label: '<', arity: 'single' },
  { value: 'isLessEqual', label: '<=', arity: 'single' },
];

const DATE: OperatorDef[] = [
  { value: 'is', label: 'is', arity: 'single' },
  { value: 'isNot', label: 'is not', arity: 'single' },
  { value: 'isBefore', label: 'is before', arity: 'single' },
  { value: 'isAfter', label: 'is after', arity: 'single' },
  { value: 'isOnOrBefore', label: 'is on or before', arity: 'single' },
  { value: 'isOnOrAfter', label: 'is on or after', arity: 'single' },
];

const SELECT: OperatorDef[] = [
  { value: 'is', label: 'is', arity: 'single' },
  { value: 'isNot', label: 'is not', arity: 'single' },
  { value: 'isAnyOf', label: 'is any of', arity: 'list' },
  { value: 'isNoneOf', label: 'is none of', arity: 'list' },
];

const MULTI_SELECT: OperatorDef[] = [
  { value: 'hasAnyOf', label: 'has any of', arity: 'list' },
  { value: 'hasAllOf', label: 'has all of', arity: 'list' },
  { value: 'hasNoneOf', label: 'has none of', arity: 'list' },
  { value: 'isExactly', label: 'is exactly', arity: 'list' },
];

const BOOL: OperatorDef[] = [{ value: 'is', label: 'is', arity: 'single' }];

const NULLABLE: OperatorDef[] = [
  { value: 'isEmpty', label: 'is empty', arity: 'none' },
  { value: 'isNotEmpty', label: 'is not empty', arity: 'none' },
];

// Map Teable field types to a logical category. Unknown types default to text.
const TYPE_CATEGORY: Record<string, FieldCategory> = {
  // text
  singleLineText: 'text',
  longText: 'text',
  formula: 'text',
  rollup: 'text',
  conditionalRollup: 'text',
  button: 'text',
  // number
  number: 'number',
  rating: 'number',
  autoNumber: 'number',
  // boolean
  checkbox: 'boolean',
  // date / time
  date: 'date',
  createdTime: 'date',
  lastModifiedTime: 'date',
  // single-value reference / choice
  singleSelect: 'select',
  user: 'select',
  link: 'select',
  createdBy: 'select',
  lastModifiedBy: 'select',
  // multi-value
  multipleSelect: 'multiSelect',
  // attachment supports emptiness checks only
  attachment: 'attachment',
};

export function categoryForType(type?: string): FieldCategory {
  if (!type) {
    return 'text';
  }
  return TYPE_CATEGORY[type] ?? 'text';
}

/** Returns the operator definitions valid for a given Teable field type. */
export function operatorsForType(type?: string): OperatorDef[] {
  switch (categoryForType(type)) {
    case 'number':
      return [...NUMBER, ...NULLABLE];
    case 'boolean':
      return [...BOOL];
    case 'date':
      return [...DATE, ...NULLABLE];
    case 'select':
      return [...SELECT, ...NULLABLE];
    case 'multiSelect':
      return [...MULTI_SELECT, ...NULLABLE];
    case 'attachment':
      return [...NULLABLE];
    case 'text':
    default:
      return [...TEXT, ...NULLABLE];
  }
}

export function operatorOptions(type?: string): Array<SelectableValue<string>> {
  return operatorsForType(type).map((o) => ({ label: o.label, value: o.value }));
}

const ALL_OPERATORS: OperatorDef[] = [
  ...TEXT,
  ...NUMBER,
  ...DATE,
  ...SELECT,
  ...MULTI_SELECT,
  ...NULLABLE,
];

export function operatorArity(op: string): 'none' | 'single' | 'list' {
  return ALL_OPERATORS.find((o) => o.value === op)?.arity ?? 'single';
}

// ---------------------------------------------------------------------------
// Template interpolation
//
// The filter object is built on the backend from the structured tree. Because
// backend data sources interpolate the query on the frontend before it is sent,
// we interpolate each condition value here so the backend receives concrete
// values.
// ---------------------------------------------------------------------------

export type ValueInterpolator = (value: string) => string;

export function interpolateFilterTree(root: FilterGroup, interpolate: ValueInterpolator): FilterGroup {
  const mapNode = (node: FilterNode): FilterNode => {
    if (node.kind === 'group') {
      return { ...node, children: node.children.map(mapNode) };
    }
    if (!node.value) {
      return node;
    }
    return { ...node, value: interpolate(node.value) };
  };
  return { ...root, children: root.children.map(mapNode) };
}

// ---------------------------------------------------------------------------
// JSON (de)serialization for persistence on the query
// ---------------------------------------------------------------------------

export function parseFilterTree(raw?: string): FilterGroup {
  if (!raw) {
    return emptyRootGroup();
  }
  try {
    const parsed = JSON.parse(raw);
    if (parsed && parsed.kind === 'group' && Array.isArray(parsed.children)) {
      return parsed as FilterGroup;
    }
  } catch {
    // ignore malformed persisted state
  }
  return emptyRootGroup();
}

export function stringifyFilterTree(root: FilterGroup): string {
  // Empty tree persists as empty string to keep queries clean.
  if (root.children.length === 0) {
    return '';
  }
  return JSON.stringify(root);
}
